staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1459 @@
+	//  - d) wait until the write deadline elapses
+	//  - e) the handler writes message 2 ("world") to the ResponseWriter object
+	//  - f) the handler invokes FlushError
+	// timeout occurs after the handler writes to and flushes the
+	// ResponseWriter object, so we expect the following:
+	//  - g) client: expected to receive a response (200 OK) from the server,
+	//       while reading the Body of the response, the client will read
+	//       message 1 from the stream, and then it will encounter an error.
+	//       the client should never see message 2.
+	//  - h) server: FlushError method of the ResponseWriter object should
+	//       return an "i/o timeout" error immediately after the write timeout
+	t.Parallel()
+
+	const deadline = 100 * time.Millisecond
+	tests := []struct {
+		protoMajor         int
+		waiter             waiter
+		clientErrVerifier  verifier
+		handlerErrVerifier verifier
+	}{
+		{
+			protoMajor: 1,
+			// write timeout is set to 100ms, a wait of 5s should
+			// be enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
+			clientErrVerifier:  wantContains{contains: "local error: tls: bad record MAC"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+		{
+			protoMajor:         2,
+			waiter:             &waitWithChannelClose{after: make(chan time.Time)},
+			clientErrVerifier:  wantContains{"stream error: stream ID 1; INTERNAL_ERROR; received from peer"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("HTTP/%d", test.protoMajor), func(t *testing.T) {
+			msg1, msg2 := "hello", "world"
+			handlerErrCh := make(chanErr, 1)
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				defer close(handlerErrCh)
+				if req.ProtoMajor != test.protoMajor {
+					t.Errorf("server: expected ProtoMajor: %d, but got: %s", test.protoMajor, req.Proto)
+					return
+				}
+				flusher, ok := w.(interface{ FlushError() error })
+				if !ok {
+					t.Errorf("server: expected ResponseWriter object to implement FlushError")
+					return
+				}
+
+				// a) write message 1
+				if _, err := w.Write([]byte(msg1)); err != nil {
+					t.Errorf("server: unexpected error from Write: %v", err)
+					return
+				}
+				// b) flush the payload in the buffer
+				if err := flusher.FlushError(); err != nil {
+					t.Errorf("server: unexpected error from FlushError: %v", err)
+					return
+				}
+
+				// c) set write deadline for this request
+				ctrl := http.NewResponseController(w)
+				if err := ctrl.SetWriteDeadline(time.Now().Add(deadline)); err != nil {
+					t.Errorf("server: unexpected error from SetWriteDeadline: %v", err)
+					return
+				}
+
+				// d) wait until the write deadline exceeds
+				<-test.waiter.wait()
+
+				// e) write message 2 (the client should never
+				//    see the following message)
```

Comment:
isn't this the actual interesting bit? Flush failing, sure. But that nothing is sent anymore.